chore: add missing course properties in IAP restore & unfulfilled events

farhan-arshad-dev opened the pull request:
### Description

- Add Properties in IAP Restore and Unfulfilled Events.

Jira: [LEARNER-10309](https://2u-internal.atlassian.net/browse/LEARNER-10309)

Commit:
chore: add missing course properties in IAP unfulfilled event

fixes: LEARNER-10309

Commit:
chore: add missing course properties in IAP restore purchases event

fixes: LEARNER-10309

HamzaIsrar12 commented on core/src/main/java/org/openedx/core/domain/interactor/IAPInteractor.kt:
```diff
@@ -109,12 +112,42 @@ class IAPInteractor(
-            purchases.filter { it.accountIdentifiers?.obfuscatedAccountId?.decodeToLong() == userId }
+        val userPurchases = purchases.filter { purchase ->
+            val userAccountId = purchase.accountIdentifiers?.obfuscatedAccountId?.decodeToLong()
+            val storeSku = purchase.accountIdentifiers?.obfuscatedProfileId?.decodeToString()
```
We can use `Purchase.getCourseSku()` extension function here.

HamzaIsrar12 commented on core/src/main/java/org/openedx/core/domain/interactor/IAPInteractor.kt:
```diff
@@ -109,12 +112,42 @@ class IAPInteractor(
         if (userPurchases.isNotEmpty()) {
-            startUnfulfilledVerification(userPurchases)
+            userPurchases.forEach { purchase ->
+                val courseVerified = enrolledCourses.find { enrolledCourse ->
+                    enrolledCourse.productInfo?.courseSku == purchase.getCourseSku()
+                }
```
This check is same as above.

farhan-arshad-dev replied:
This is intended to trace the course from the `enrolledCourses` list, allowing us to proceed with the Unfulfilled process.

HamzaIsrar12 commented on core/src/main/java/org/openedx/core/system/notifier/app/AppNotifier.kt:
```diff
@@ -16,4 +21,7 @@ class AppNotifier {
 
     suspend fun send(event: AppUpgradeEvent) = channel.emit(event)
 
+    suspend fun send(event: EnrolledCourseEvent) = channel.emit(event)
+
+    suspend fun send(event: RequestEnrolledCourseEvent) = channel.emit(event)
```
I think we should rename them and place them in `IAPNotifier`.

farhan-arshad-dev replied:
I have created a common event to fetch the `EnrolledCourse`, making it reusable in other areas if needed.

HamzaIsrar12 reviewed (changes requested):
We have four additional improvements to address that I forgot to mention in the underline ticket:  

- Ensure the unfulfilled flow functions correctly, even when there's only a single course.  
- Trigger `CourseUpgradeSuccess` regardless of the user's choice on the Unfulfilled Refresh-or-Continue dialog.  
- Include these properties in the `CourseUpgradeSuccess` event during the unfulfilled flow.  
- Since the flow sends only one event for all unfulfilled courses, it's better to limit it to one process per session or click to ensure the correct number of `CourseUpgradeSuccess` events are sent.  

Thank you!

Commit:
fix: address PR comments

farhan-arshad-dev commented:
> Ensure the unfulfilled flow functions correctly, even when there's only a single course.

done.

> Trigger CourseUpgradeSuccess regardless of the user's choice on the Unfulfilled Refresh-or-Continue dialog.

done.

> Include these properties in the CourseUpgradeSuccess event during the unfulfilled flow.

done.

> Since the flow sends only one event for all unfulfilled courses, it's better to limit it to one process per session or click to ensure the correct number of CourseUpgradeSuccess events are sent.

done.

HamzaIsrar12 commented on core/src/main/java/org/openedx/core/domain/interactor/IAPInteractor.kt:
```diff
@@ -109,52 +112,83 @@ class IAPInteractor(
         if (userPurchases.isNotEmpty()) {
-            startUnfulfilledVerification(userPurchases)
-            return true
+            userPurchases[0].let { purchase ->
```
Please use `.first()` instead.

farhan-arshad-dev replied:
java things 😄 .

HamzaIsrar12 commented on core/src/main/java/org/openedx/core/domain/interactor/IAPInteractor.kt:
```diff
@@ -109,52 +112,83 @@ class IAPInteractor(
+            }
         } else {
-            purchases.forEach {
+            purchases.subtract(userPurchases.toSet()).forEach {
```
The `userPurchases` will always be empty.

HamzaIsrar12 commented on course/src/main/java/org/openedx/course/presentation/container/CourseContainerViewModel.kt:
```diff
@@ -286,7 +286,6 @@ class CourseContainerViewModel(
                         }
                         if (isIAPFlow) {
                             if (isExpiredCoursePurchase) {
-                                eventLogger.upgradeSuccessEvent()
```
Is there a specific reason we removed this? Removing it would prevent the Upgrade Success event from being triggered in the Expired Course scenario.

farhan-arshad-dev replied:
reverted.

HamzaIsrar12 commented on core/src/main/java/org/openedx/core/presentation/iap/IAPViewModel.kt:
```diff
@@ -84,7 +84,9 @@ class IAPViewModel(
                 when (event) {
                     is CourseDataUpdated -> {
-                        eventLogger.upgradeSuccessEvent()
+                        if (eventLogger.isSilentIAPFlow == null) {
```
You can use `isNull` instead.

HamzaIsrar12 reviewed (changes requested):
- The `Refresh Now` sends another `Course Upgrade Success` event
- No properties available in `Course Upgrade Success` after Restore Purchase

Commit:
fix: address PR comments-2

farhan-arshad-dev commented:
> The Refresh Now sends another Course Upgrade Success event

fixed.

> No properties available in Course Upgrade Success after Restore Purchase

added.

Commit:
fix: black screen issue while refreshing the course data

k1rill reviewed (approved):
LGTM

HamzaIsrar12 reviewed (approved):
I tested the following flows, and the `Payment:Course Upgrade Success` event is firing as expected with all the necessary properties needed in Analytics:

**Tested Flows:**  
- Restore Mode  
- Silent Mode  
- Enrollments Screen  
- Dashboard Screen  
- Expired Screen
